### PR TITLE
[YUNIKORN-14] Add rest API to retrieve app/container history info

### PR DESCRIPTION
**Pull request by adamantal**
This PR only handles the server (core) part of this issue. The UI related changes will be handled in [YUNIKORN-8](https://issues.apache.org/jira/browse/YUNIKORN-8).

The initial implementation consists of creating a `HistoricalPartitionInfo` object which stores some metrics (currently the number of application and containers). This is updated in 10 minutes by `HistoricalPartitionInfoUpdater`

If there's need the poll time can be configurable, right now it is just hardcoded to 10 minutes, and we will keep 1 day of history (144 items).

### CODE REVIEW

File: pkg/webservice/handlers.go [hunk 1]
```diff
@@ -235,3 +236,57 @@ func getNodeJSON(nodeInfo *cache.NodeInfo) *dao.NodeDAOInfo {
 		Schedulable: nodeInfo.IsSchedulable(),
 	}
 }
+
+func GetApplicationHistory(w http.ResponseWriter, r *http.Request) {
+	writeHeaders(w)
+
+	var result []*dao.ApplicationHistoryDAOInfo
+	lists := gClusterInfo.ListPartitions()
+	for _, k := range lists {
```
Comment (yangwwei) [inline pkg/webservice/handlers.go]: right now it reports back per partition
I don't think front-end expect this. we should aggregate this to the cluster level. Can we change the `HistoricPartitionInfo` to `HistoricClusterInfo`?

File: pkg/cache/history/history_info.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+package history
+
+import (
+	"time"
+)
+
+var LIMIT = 144 // 10 minutes in 1 day
+
+type HistoricalPartitionInfo struct {
+	records []*PartitionInfoRecord
```
[the remark this answers is not in the document]
Comment (adamantal) [reply pkg/cache/history/history_info.go]: Added a simple `sync.Mutex` for this.

File: pkg/cache/history/history_info.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+type PartitionInfoRecord struct {
+	Timestamp         time.Time
+	TotalApplications int
+	TotalContainers   int
+}
+
+func (hpInfo *HistoricalPartitionInfo) Store(totalApplications, totalContainers int) {
```
[the remark this answers is not in the document]
Comment (adamantal) [reply pkg/cache/history/history_info.go]: Done (for `HistoricalClusterInfo`)

File: pkg/cache/history/history_info.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+ limitations under the License.
+*/
+
+package history
+
+import (
+	"time"
+)
+
+var LIMIT = 144 // 10 minutes in 1 day
```
[the remark this answers is not in the document]
Comment (adamantal) [reply pkg/cache/history/history_info.go]: Done

[the remark this answers is not in the document]
Comment (adamantal) [reply pkg/cache/history/history_info.go]: Modified the constants accordingly

File: pkg/cache/history_collector.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+package cache
+
+import (
+	"time"
+
+	"github.com/apache/incubator-yunikorn-core/pkg/cache/history"
+	"github.com/apache/incubator-yunikorn-core/pkg/log"
+)
+
+var tickerDefault = 10 * time.Minute
```
[the remark this answers is not in the document]
Comment (adamantal) [reply pkg/cache/history_collector.go]: Done

File: pkg/cache/history_collector.go [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+func (hpInfoUpdater *HistoricalPartitionInfoUpdater) Start() {
+	go func() {
+		for {
+			select {
+			case <-hpInfoUpdater.finished:
+				return
+			case <-hpInfoUpdater.ticker.C:
+				log.Logger().Info("Adding current status to historical partition data")
```
[the remark this answers is not in the document]
Comment (adamantal) [reply pkg/cache/history_collector.go]: Done

File: pkg/cache/partition_info.go [hunk 7]
```diff
@@ -103,6 +106,11 @@ func newPartitionInfo(partition configs.PartitionConfig, rmID string, info *Clus
 	// TODO get the resolver from the config
 	p.userGroupCache = security.GetUserGroupCache("")
 
+	hpInfo := &history.HistoricalPartitionInfo{}
+	p.history = hpInfo
+	p.historyUpdater = NewHistoricalPartitionInfoUpdater(p, hpInfo)
+	p.historyUpdater.Start()
+
```
[the remark this answers is not in the document]
Comment (adamantal) [reply pkg/cache/partition_info.go]: I only moved the `Start` call to `StartService`, but I kept the construction of the `HistoricalParitionInfoUpdater` object in the constructor.

File: pkg/webservice/handlers.go (hunk 1 above)
Comment (adamantal) [reply pkg/webservice/handlers.go]: Agreed. Modified this and other classes in the PR accordingly.

**Comment by adamantal** [comment]
Strangely the test passes on my local - I didn't touch `scheduler_recovery_test.go`, so I think it is probably unrelated.

**Comment by adamantal** [comment]
Tested the patch locally with the following results:
```
AAntal-MBP15:yunikorn-core adamantal$ kubectl logs yunikorn-scheduler-76ff98dff4-lb77q yunikorn-scheduler-k8s | grep "Adding current status"
2020-03-26T15:35:10.911Z	DEBUG	cache/history_collector.go:61	Adding current status to historical partition data
2020-03-26T15:36:10.912Z	DEBUG	cache/history_collector.go:61	Adding current status to historical partition data
2020-03-26T15:37:10.912Z	DEBUG	cache/history_collector.go:61	Adding current status to historical partition data
```
So as expected, we are collecting the metrics in every minute.

I also checked the rest API endpoint, and it seems like there's a bug. I missed a "/" from the beginning of the REST endpoints, so they are not accessible. I will fix it, test it locally and update the PR.

**Comment by adamantal** [comment]
Fixed the bug and tried out locally.

Sample output of `localhost:9080/ws/v1/history/containers`:
```json
[
    {
        "timestamp": 1585238263802408000,
        "totalContainers": "0"
    },
    {
        "timestamp": 1585238323803183000,
        "totalContainers": "0"
    },
    {
        "timestamp": 1585238383803268000,
        "totalContainers": "0"
    },
    {
        "timestamp": 1585238443804046000,
        "totalContainers": "0"
    },
    {
        "timestamp": 1585238503806253000,
        "totalContainers": "0"
    },
    {
        "timestamp": 1585238563807452000,
        "totalContainers": "0"
    }
]
```

**Comment by adamantal** [comment]
Created a brand new architecture according to @wilfred-s comment on the jira.

Checked on my local with the `make run` command on k8shim, also starting 3 sleep pods using the `deployments/examples/sleep/sleeppods.yaml` file.
Example output of the `(localhost:9080)/ws/v1/history/containers` endpoint:

```json
[
    {
        "timestamp": 1585579879950325000,
        "totalContainers": "0"
    },
    {
        "timestamp": 1585579939947939000,
        "totalContainers": "0"
    },
    {
        "timestamp": 1585579999942027000,
        "totalContainers": "0"
    },
    {
        "timestamp": 1585580059941905000,
        "totalContainers": "3"
    }
]
```

**Comment by yangwwei** [comment]
Thanks @adamantal !
@wilfred-s , can you help to review this change?

**Comment by adamantal** [comment]
Rebased due to conflict in the `go.mod` file (YUNIKORN-63).

File: pkg/webservice/handlers.go [hunk 8]
```diff
@@ -236,3 +236,39 @@ func getNodeJSON(nodeInfo *cache.NodeInfo) *dao.NodeDAOInfo {
 		Schedulable: nodeInfo.IsSchedulable(),
 	}
 }
+
+func GetApplicationHistory(w http.ResponseWriter, r *http.Request) {
+	writeHeaders(w)
+
+	var result []*dao.ApplicationHistoryDAOInfo
+	records := imHistory.GetRecords()
+	for _, record := range records {
```
Comment (wilfred-s) [inline pkg/webservice/handlers.go]: Need to check the impact of the earlier mentioned unlimited growth solution on this iteration loop

File: pkg/metrics/metrics_collector_test.go [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+	metricsCollector.StartService()
+
+	go func() {
+		metrics := GetSchedulerMetrics()
+		i := 0
+		for i < 3 {
+			metrics.IncTotalApplicationsRunning()
+			metrics.AddAllocatedContainers(2)
+			i += 1
+			time.Sleep(400 * time.Millisecond)
```
Comment (wilfred-s) [inline pkg/metrics/metrics_collector_test.go]: Do we really need to sleep between each update? Extends the test run unneeded

File: pkg/metrics/metrics_collector_test.go [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+	"time"
+
+	"gotest.tools/assert"
+
+	"github.com/apache/incubator-yunikorn-core/pkg/metrics/history"
+)
+
+func TestHistoricalPartitionInfoUpdater(t *testing.T) {
+	metricsHistory := history.NewInternalMetricsHistory(3)
+	setInternalMetricsCollectorTickerForTest(500 * time.Millisecond)
```
Comment (wilfred-s) [inline pkg/metrics/metrics_collector_test.go]: Can we shorten this to something like 10 msec?

File: pkg/metrics/metrics_collector.go [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+			}
+		}
+	}()
+}
+
+func (u *InternalMetricsCollector) Stop() {
+	u.stopped <- true
+}
+
+func setInternalMetricsCollectorTickerForTest(newDefault time.Duration) {
```
Comment (wilfred-s) [inline pkg/metrics/metrics_collector.go]: Use a comment like `visible for test only` or something like it
We should not use _test_ in the function name unless it is really a test.

File: pkg/metrics/metrics_collector.go [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+	tickerDefault = newDefault
+}
```
Comment (wilfred-s) [inline pkg/metrics/metrics_collector.go]: missing newline

File: pkg/metrics/init.go [hunk 13]
```diff
@@ -54,6 +56,7 @@ type CoreSchedulerMetrics interface {
 	// Metrics Ops related to ScheduledAllocationSuccesses
 	IncAllocatedContainer()
 	AddAllocatedContainers(value int)
+	getAllocatedContainers() *prometheus.Counter
```
Comment (wilfred-s) [inline pkg/metrics/init.go]: can we not simply return the primitive we need?
Hide all the prometheus logic inside one file, see other review comments on how.

Comment (adamantal) [reply pkg/metrics/init.go]: Makes sense, fixed

File: pkg/metrics/metrics_collector_test.go (hunk 9 above)
Comment (adamantal) [reply pkg/metrics/metrics_collector_test.go]: I wanted to wait for at least 2 cycles. Do you think it is still an issue after ticker has been configured to 10msec (from comment below)?

File: pkg/metrics/history/internal_metrics.go [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+		records: make([]*MetricsRecord, 0),
+		limit:   limit,
+	}
+}
+
+func (h *InternalMetricsHistory) Store(totalApplications, totalContainers int) {
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
+
+	h.records = append(h.records,
```
[the remark this answers is not in the document]
Comment (adamantal) [reply pkg/metrics/history/internal_metrics.go]: Nice catch. I actually got through that post during my Go tutorial, but didn't realize that I made a mistake here until now.
I started with a containers/list implementation, but got into problems, because it was not typesafe and had to case everywhere where it is used.

Implemented a circular, sized array using the pointer and the limit variables. If there's any official typesafe implementation in Golang, or some popular package to carry out this, please let me know and I'll use that. I think this is the most light implementation that we could use.

File: go.mod [hunk 15]
```diff
@@ -4,7 +4,7 @@ go 1.12
 
 require (
 	github.com/apache/incubator-yunikorn-scheduler-interface v0.0.0-20200327234544-149aaa3d8e48
-	github.com/golang/protobuf v1.3.5 // indirect
```
Comment (sunilgovind) [inline go.mod]: could we avoid this change

File: pkg/metrics/history/internal_metrics.go [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+		for _, record := range h.records {
+			if record != nil {
+				returnRecords = append(returnRecords, record)
+			}
+		}
+	}
+	return returnRecords
+}
+
+func (h *InternalMetricsHistory) GetLimit() int {
```
Comment (sunilgovind) [inline pkg/metrics/history/internal_metrics.go]: Limit is set only one time when system is booted up, could you please confirm on that.
If that's the case, do we need a mutex here?

File: pkg/metrics/metrics_collector.go [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+				log.Logger().Debug("Adding current status to historical partition data")
+
+				totalAppsRunning, err := m.scheduler.getTotalApplicationsRunning()
+				if err != nil {
+					log.Logger().Warn("Could not encode metric.", zap.Error(err))
+					continue
+				}
+				totalContainersRunning, err := m.scheduler.getAllocatedContainers()
+				if err != nil {
+					log.Logger().Warn("Could not encode metric.", zap.Error(err))
```
Comment (sunilgovind) [inline pkg/metrics/metrics_collector.go]: cud we add metric name also here?

File: pkg/webservice/handlers.go [hunk 18]
```diff
@@ -236,3 +236,39 @@ func getNodeJSON(nodeInfo *cache.NodeInfo) *dao.NodeDAOInfo {
...
+		element := &dao.ApplicationHistoryDAOInfo{
+			Timestamp:         record.Timestamp.UnixNano(),
+			TotalApplications: strconv.Itoa(record.TotalApplications),
```
Comment (sunilgovind) [inline pkg/webservice/handlers.go]: in general, please check spacing. is it tab or spaces?
code allignment?

File: pkg/webservice/handlers.go [hunk 19]
```diff
@@ -236,3 +236,39 @@ func getNodeJSON(nodeInfo *cache.NodeInfo) *dao.NodeDAOInfo {
...
+
+func GetContainerHistory(w http.ResponseWriter, r *http.Request) {
+	writeHeaders(w)
+
+	var result []*dao.ContainerHistoryDAOInfo
+	records := imHistory.GetRecords()
+	for _, record := range records {
+		element := &dao.ContainerHistoryDAOInfo{
+			Timestamp:       record.Timestamp.UnixNano(),
+			TotalContainers: strconv.Itoa(record.TotalContainers),
```
Comment (sunilgovind) [inline pkg/webservice/handlers.go]: are we having only TotalContainers/Apps returned in an array when we query /ws/v1/history/containers ?
Cud u pls share a sample response?

File: pkg/metrics/history/internal_metrics.go [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+	limit   int
+
+	// internal implementation of limited array
+	pointer int
+	full    bool
+
+	sync.RWMutex
+}
+
+type MetricsRecord struct {
```
Comment (wilfred-s) [inline pkg/metrics/history/internal_metrics.go:40]: Do we need to export this?
IDE says there is no need to export.

File: pkg/metrics/metrics_collector.go [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+
+	"go.uber.org/zap"
+
+	"github.com/apache/incubator-yunikorn-core/pkg/log"
+	"github.com/apache/incubator-yunikorn-core/pkg/metrics/history"
+)
+
+var tickerDefault = 1 * time.Minute
+
+type InternalMetricsCollector struct {
```
Comment (wilfred-s) [inline pkg/metrics/metrics_collector.go:32]: Do we need to export?
IDE says that it is not needed

File: go.mod (hunk 15 above)
Comment (wilfred-s) [reply go.mod:7]: a `go mod tidy` adds it back again on my system so I think we need to follow that and remove this from the change

File: pkg/webservice/handlers.go (hunk 18 above)
Comment (adamantal) [reply pkg/webservice/handlers.go:248]: What is the intended usage of the `go fmt` command?

I have the following error message running it from the project's root:
```
can't load package: package github.com/apache/incubator-yunikorn-core: unknown import path "github.com/apache/incubator-yunikorn-core": cannot find module providing package github.com/apache/incubator-yunikorn-core
```

File: pkg/entrypoint/entrypoint.go [hunk 22]
```diff
@@ -73,14 +73,19 @@ func startAllServicesWithParameters(opts StartupOptions) *ServiceContext {
...
-	metricsCollector.StartService()
 
 	context := &ServiceContext{
 		RMProxy:   proxy,
 		Cache:     cache,
 		Scheduler: scheduler,
 	}
 
+	imHistory := history.NewInternalMetricsHistory(opts.metricsHistorySize)
+	if opts.metricsHistorySize != 0 {
```
[the remark this answers is not in the document]
Comment (adamantal) [reply pkg/entrypoint/entrypoint.go:84]: Agreed. Fixed

File: pkg/metrics/history/internal_metrics.go (hunk 20 above)
Comment (adamantal) [reply pkg/metrics/history/internal_metrics.go:40]: It is used in `webservice/handlers.go`, though the name `MetricsRecord` is not explicitly used. I fix it then.

File: pkg/metrics/history/internal_metrics.go [hunk 23]
```diff
@@ -40,35 +45,55 @@ type MetricsRecord struct {
...
+				pointer = 0
+			}
+		}
+	} else {
+		for _, record := range h.records {
+			if record != nil {
+				returnRecords = append(returnRecords, record)
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (adamantal) [reply pkg/metrics/history/internal_metrics.go:91]: Indeed, I fixed accordingly.

I'd be more comfortable with filtering the nil values here, assuming we could have other caller's of this function, but right now it is ok to do it in the webapp.

File: pkg/metrics/metrics_collector_test.go (hunk 9 above)
Comment (wilfred-s) [reply pkg/metrics/metrics_collector_test.go]: with a shorter sleep I am OK with the test and the wait we put in.

File: pkg/metrics/history/internal_metrics.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (wilfred-s) [reply pkg/metrics/history/internal_metrics.go]: new impl looks good

File: pkg/webservice/handlers.go (hunk 8 above)
Comment (adamantal) [reply pkg/webservice/handlers.go]: Could you check whether the fixing commit is good? I hope I get your first item right.

File: pkg/metrics/scheduler.go [hunk 24]
```diff
@@ -200,6 +201,14 @@ func (m *SchedulerMetrics) AddAllocatedContainers(value int) {
 	m.allocatedContainers.Add(float64(value))
 }
 
+func (m *SchedulerMetrics) getAllocatedContainers() (value int, err error) {
```
Comment (wilfred-s) [inline pkg/metrics/scheduler.go:204]: We should not mix named and unnamed returns.
We have standardised on unnamed everywhere else.

File: pkg/metrics/scheduler.go [hunk 25]
```diff
@@ -265,6 +274,12 @@ func (m *SchedulerMetrics) SetTotalApplicationsRunning(value int) {
 	m.totalApplicationsRunning.Set(float64(value))
 }
 
+func (m *SchedulerMetrics) getTotalApplicationsRunning() (int, error) {
+	metric := &dto.Metric{}
+	err := m.totalApplicationsRunning.Write(metric)
+	return int(*metric.Gauge.Value), err
+}
```
Comment (wilfred-s) [inline pkg/metrics/scheduler.go:281]: See `getAllocatedContainers` same nil pointer here, solution is same, don't use named returns

File: pkg/webservice/handlers.go [hunk 26]
```diff
@@ -236,3 +236,49 @@ func getNodeJSON(nodeInfo *cache.NodeInfo) *dao.NodeDAOInfo {
...
 }
+
+func GetApplicationHistory(w http.ResponseWriter, r *http.Request) {
+	writeHeaders(w)
+
+	if imHistory != nil {
+		var result []*dao.ApplicationHistoryDAOInfo
+		records := imHistory.GetRecords()
+		for _, record := range records {
+			if record != nil {
```
Comment (wilfred-s) [inline pkg/webservice/handlers.go:247]: Can we have records that are not nil after the first nil record? I think not, but please check.
If we cannot why not use a break to jump out of the loop:
```
if record == nil {
	break
}
```

File: pkg/webservice/handlers.go [hunk 27]
```diff
@@ -236,3 +236,49 @@ func getNodeJSON(nodeInfo *cache.NodeInfo) *dao.NodeDAOInfo {
 		Schedulable: nodeInfo.IsSchedulable(),
 	}
 }
+
+func GetApplicationHistory(w http.ResponseWriter, r *http.Request) {
+	writeHeaders(w)
+
+	if imHistory != nil {
```
Comment (wilfred-s) [inline pkg/webservice/handlers.go:243]: use unindented flow instead of an else clause:
```
if imHistory == nil {
	http.Error...
	return
}
var result ....
```

File: pkg/webservice/handlers.go [hunk 28]
```diff
@@ -236,3 +236,49 @@ func getNodeJSON(nodeInfo *cache.NodeInfo) *dao.NodeDAOInfo {
...
+		}
+		if err := json.NewEncoder(w).Encode(result); err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		}
+	} else {
+		http.Error(w, "Internal metrics collection is not enabled.", http.StatusInternalServerError)
+	}
+}
+
+func GetContainerHistory(w http.ResponseWriter, r *http.Request) {
```
Comment (wilfred-s) [inline pkg/webservice/handlers.go:263]: same comments as per `GetApplicationHistory`

File: pkg/metrics/metrics_collector_test.go [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+		}
+	}()
+
+	time.Sleep(11 * time.Millisecond)
+	metricsCollector.Stop()
+
+	records := metricsHistory.GetRecords()
+	assert.Equal(t, 3, len(records), "Expected exactly 2 history records")
+	for i, record := range records {
+		if i == 0 {
```
Comment (wilfred-s) [inline pkg/metrics/metrics_collector_test.go:52]: use a switch statement or just direct `records[0]` etc

File: pkg/metrics/metrics_collector_test.go [hunk 30]
```diff
@@ -0,0 +1,62 @@
...
+			i += 1
+			time.Sleep(4 * time.Millisecond)
+		}
+	}()
+
+	time.Sleep(11 * time.Millisecond)
+	metricsCollector.Stop()
+
+	records := metricsHistory.GetRecords()
+	assert.Equal(t, 3, len(records), "Expected exactly 2 history records")
```
Comment (wilfred-s) [inline pkg/metrics/metrics_collector_test.go:50]: nit: text not updated to 3

Comment (adamantal) [reply pkg/metrics/metrics_collector_test.go:50]: Silly of me, thanks for reminding!

File: pkg/webservice/handlers.go (hunk 26 above)
Comment (adamantal) [reply pkg/webservice/handlers.go:247]: Fixed this, but I think we're relying too much on the actual implementation of the `internalMetricsHistory`, and it makes the code too fragile if it is modified in the future.

File: pkg/metrics/scheduler.go (hunk 24 above)
Comment (adamantal) [reply pkg/metrics/scheduler.go:204]: Thanks, didn't know that. Is it standardised on a per file or a per package base?

File: pkg/metrics/metrics_collector.go [hunk 31]
```diff
@@ -0,0 +1,81 @@
...
+			select {
+			case <-u.stopped:
+				return
+			case <-u.ticker.C:
+				log.Logger().Debug("Adding current status to historical partition data")
+
+				totalAppsRunning, err := m.scheduler.getTotalApplicationsRunning()
+				if err != nil {
+					log.Logger().Warn("Could not encode totalApplications metric.", zap.Error(err))
+					continue
```
[the remark this answers is not in the document]
Comment (adamantal) [reply pkg/metrics/metrics_collector.go:61]: I used -1, added a comment on the class.

**Comment by yangwwei** [comment]
Looks good to me, thanks @wilfred-s , @adamantal !
I think we can merge this. Only one minor thing, we might want to check with the front-end. I saw when we are not able to get app/container count, it is set to `-1`.  We need to make sure the front-end supports this value.

**Comment by wilfred-s** [comment]
Thank you Weiwei, I'll log a jira against the webapp to make sure we can handle the -1 value.
